File: predicate-search/src/test/java/com/yahoo/search/predicate/annotator/PredicateTreeAnalyzerTest.java
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.search.predicate.annotator;

import com.yahoo.document.predicate.FeatureConjunction;
import com.yahoo.document.predicate.Predicate;
import com.yahoo.document.predicate.PredicateOperator;
import org.junit.jupiter.api.Test;

import java.util.Arrays;

import static com.yahoo.document.predicate.Predicates.and;
import static com.yahoo.document.predicate.Predicates.feature;
import static com.yahoo.document.predicate.Predicates.not;
import static com.yahoo.document.predicate.Predicates.or;
import static org.junit.jupiter.api.Assertions.*;

public class PredicateTreeAnalyzerTest {

    @Test
    void require_that_minfeature_is_1_for_simple_term() {
        Predicate p = feature("foo").inSet("bar");
        PredicateTreeAnalyzerResult r = PredicateTreeAnalyzer.analyzePredicateTree(p);
        assertEquals(1, r.minFeature);
        assertEquals(1, r.treeSize);
        assertTrue(r.sizeMap.isEmpty());
    }

    @Test
    void require_that_minfeature_is_1_for_simple_negative_term() {
        Predicate p = not(feature("foo").inSet("bar"));
        PredicateTreeAnalyzerResult r = PredicateTreeAnalyzer.analyzePredicateTree(p);
        assertEquals(1, r.minFeature);
    }

    @Test
    void require_that_minfeature_is_sum_for_and() {
        Predicate p =
                and(
                        feature("foo").inSet("bar"),
                        feature("baz").inSet("qux"),
                        feature("quux").inSet("corge"));
        PredicateTreeAnalyzerResult r = PredicateTreeAnalyzer.analyzePredicateTree(p);
        assertEquals(3, r.minFeature);
        assertEquals(3, r.treeSize);
        assertEquals(3, r.sizeMap.size());
        assertSizeMapContains(r, pred(p).child(0), 1);
        assertSizeMapContains(r, pred(p).child(1), 1);
        assertSizeMapContains(r, pred(p).child(2), 1);
    }

    @Test
    void require_that_minfeature_is_min_for_or() {
        Predicate p =
                or(
                        and(
                                feature("foo").inSet("bar"),
                                feature("baz").inSet("qux"),
                                feature("quux").inSet("corge")),
                        and(
                                feature("grault").inSet("garply"),
                                feature("waldo").inSet("fred")));
        PredicateTreeAnalyzerResult r = PredicateTreeAnalyzer.analyzePredicateTree(p);
        assertEquals(2, r.minFeature);
        assertEquals(5, r.treeSize);
        assertEquals(5, r.sizeMap.size());
        assertSizeMapContains(r, pred(p).child(0).child(0), 1);
        assertSizeMapContains(r, pred(p).child(0).child(1), 1);
        assertSizeMapContains(r, pred(p).child(0).child(2), 1);
        assertSizeMapContains(r, pred(p).child(1).child(0), 1);
        assertSizeMapContains(r, pred(p).child(1).child(1), 1);
    }

    @Test
    void require_that_minfeature_rounds_up() {
        Predicate p =
                or(
                        feature("foo").inSet("bar"),
                        feature("foo").inSet("bar"),
                        feature("foo").inSet("bar"));
        PredicateTreeAnalyzerResult r = PredicateTreeAnalyzer.analyzePredicateTree(p);
        assertEquals(1, r.minFeature);
        assertEquals(3, r.treeSize);
    }

    @Test
    void require_that_minvalue_feature_set_considers_all_values() {
        {
            Predicate p =
                    and(
                            feature("foo").inSet("A", "B"),
                            feature("foo").inSet("B"));
            PredicateTreeAnalyzerResult r = PredicateTreeAnalyzer.analyzePredicateTree(p);
            assertEquals(1, r.minFeature);
            assertEquals(2, r.treeSize);
        }
        {
            Predicate p =
                    and(
                            feature("foo").inSet("A", "B"),
                            feature("foo").inSet("C"));
            PredicateTreeAnalyzerResult r = PredicateTreeAnalyzer.analyzePredicateTree(p);
            assertEquals(2, r.minFeature);
            assertEquals(2, r.treeSize);
        }
    }

    @Test
    void require_that_not_features_dont_count_towards_minfeature_calculation() {
        Predicate p =
                and(
                        feature("foo").inSet("A"),
                        not(feature("foo").inSet("A")),
                        not(feature("foo").inSet("B")),
                        feature("foo").inSet("B"));
        PredicateTreeAnalyzerResult r = PredicateTreeAnalyzer.analyzePredicateTree(p);
        assertEquals(3, r.minFeature);
        assertEquals(6, r.treeSize);
    }

    @Test
    void require_that_multilevel_and_stores_size() {
        Predicate p =
                and(
                        and(
                                feature("foo").inSet("bar"),
                                feature("baz").inSet("qux"),
                                feature("quux").inSet("corge")),
                        and(
                                feature("grault").inSet("garply"),
                                feature("waldo").inSet("fred")));
        PredicateTreeAnalyzerResult r = PredicateTreeAnalyzer.analyzePredicateTree(p);
        assertEquals(5, r.minFeature);
        assertEquals(5, r.treeSize);
        assertEquals(7, r.sizeMap.size());
        assertSizeMapContains(r, pred(p).child(0), 3);
        assertSizeMapContains(r, pred(p).child(1), 2);
        assertSizeMapContains(r, pred(p).child(0).child(0), 1);
        assertSizeMapContains(r, pred(p).child(0).child(1), 1);
        assertSizeMapContains(r, pred(p).child(0).child(2), 1);
        assertSizeMapContains(r, pred(p).child(1).child(0), 1);
        assertSizeMapContains(r, pred(p).child(1).child(1), 1);
    }

    @Test
    void require_that_not_ranges_dont_count_towards_minfeature_calculation() {
        Predicate p =
                and(
                        feature("foo").inRange(0, 10),
                        not(feature("foo").inRange(0, 10)),
                        feature("bar").inRange(0, 10),
                        not(feature("bar").inRange(0, 10)));
        PredicateTreeAnalyzerResult r = PredicateTreeAnalyzer.analyzePredicateTree(p);
        assertEquals(3, r.minFeature);
        assertEquals(6, r.treeSize);
    }

    @Test
    void require_that_featureconjunctions_contribute_as_one_feature() {
        Predicate p =
                conj(
                        feature("foo").inSet("bar"),
                        feature("baz").inSet("qux"));
        PredicateTreeAnalyzerResult r = PredicateTreeAnalyzer.analyzePredicateTree(p);
        assertEquals(1, r.minFeature);
        assertEquals(1, r.treeSize);
    }

    @Test
    void require_that_featureconjunctions_count_as_leaf_in_subtree_calculation() {
        Predicate p =
                and(
                        and(
                                feature("grault").inRange(0, 10),
                                feature("waldo").inRange(0, 10)),
                        conj(
                                feature("foo").inSet("bar"),
                                feature("baz").inSet("qux"),
                                feature("quux").inSet("corge")));
        PredicateTreeAnalyzerResult r = PredicateTreeAnalyzer.analyzePredicateTree(p);
        assertEquals(3, r.minFeature);
        assertEquals(3, r.treeSize);
        assertEquals(4, r.sizeMap.size());
        assertSizeMapContains(r, pred(p).child(0), 2);
        assertSizeMapContains(r, pred(p).child(0).child(0), 1);
        assertSizeMapContains(r, pred(p).child(0).child(1), 1);
        assertSizeMapContains(r, pred(p).child(1), 1);
    }

    @Test
    void require_that_multiple_indentical_feature_conjunctions_does_not_contribute_more_than_one() {
        Predicate p =
                and(
                        or(
                                conj(
                                        feature("a").inSet("b"),
                                        feature("c").inSet("d")
                                ),
                                feature("x").inSet("y")),
                        or(
                                conj(
                                        feature("a").inSet("b"),
                                        feature("c").inSet("d")
                                ),
                                feature("z").inSet("w")));
        PredicateTreeAnalyzerResult r = PredicateTreeAnalyzer.analyzePredicateTree(p);
        assertEquals(1, r.minFeature);
        assertEquals(4, r.treeSize);
    }

    private static FeatureConjunction conj(Predicate... operands) {
        return new FeatureConjunction(Arrays.asList(operands));
    }

    private static void assertSizeMapContains(PredicateTreeAnalyzerResult r, PredicateSelector selector, int expectedValue) {
        Integer actualValue = r.sizeMap.get(selector.predicate);
        assertNotNull(actualValue);
        assertEquals(expectedValue, actualValue.intValue());
    }

    private static class PredicateSelector {
        public final Predicate predicate;

        public PredicateSelector(Predicate predicate) {
            this.predicate = predicate;
        }

        public PredicateSelector child(int index) {
            PredicateOperator op = (PredicateOperator) predicate;
            return new PredicateSelector(op.getOperands().get(index));
        }
    }

    private static PredicateSelector pred(Predicate p) {
        return new PredicateSelector(p);
    }
}

File: searchlib/src/tests/predicate/predicate_tree_analyzer_test.cpp
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
// Unit tests for PredicateTreeAnalyzer.

#include <vespa/document/predicate/predicate.h>
#include <vespa/document/predicate/predicate_slime_builder.h>
#include <vespa/searchlib/predicate/predicate_tree_analyzer.h>
#include <vespa/vespalib/gtest/gtest.h>
#include <iomanip>
#include <sstream>

using document::PredicateSlimeBuilder;
using namespace search;
using namespace search::predicate;
using document::Predicate;
using vespalib::Slime;
using vespalib::slime::Cursor;
using std::map;
using std::string;

namespace {
using Builder = PredicateSlimeBuilder;

TEST(PredicateTreAnalyzerTest, require_that_minfeature_is_1_for_simple_term) {
    auto slime(Builder().feature("foo").value("bar").build());
    PredicateTreeAnalyzer analyzer(slime->get());
    EXPECT_EQ(1, analyzer.getMinFeature());
    EXPECT_EQ(1, analyzer.getSize());
    EXPECT_TRUE(analyzer.getSizeMap().empty());
}

TEST(PredicateTreAnalyzerTest, require_that_minfeature_is_1_for_simple_negative_term) {
    auto slime(Builder().neg().feature("foo").value("bar").build());
    PredicateTreeAnalyzer analyzer(slime->get());
    EXPECT_EQ(1, analyzer.getMinFeature());
    EXPECT_EQ(2, analyzer.getSize());
}

void checkSizeMap(const map<string, int> &map, const string &key, int val) {
    std::ostringstream os;
    os << "key=" << std::quoted(key) << ", val=" << val;
    SCOPED_TRACE(os.str());
    auto it = map.find(key);
    ASSERT_TRUE(it != map.end());
    EXPECT_EQ(val, it->second);
}

TEST(PredicateTreAnalyzerTest, require_that_minfeature_is_sum_for_and) {
    auto slime(Builder()
               .and_node({Builder().feature("foo").value("bar"),
                          Builder().feature("baz").value("qux"),
                          Builder().feature("quux").value("corge")}).build());
    PredicateTreeAnalyzer analyzer(slime->get());
    EXPECT_EQ(3, analyzer.getMinFeature());
    EXPECT_EQ(3, analyzer.getSize());
    EXPECT_EQ(3u, analyzer.getSizeMap().size());
    checkSizeMap(analyzer.getSizeMap(), "a0", 1);
    checkSizeMap(analyzer.getSizeMap(), "a1", 1);
    checkSizeMap(analyzer.getSizeMap(), "a2", 1);
}

TEST(PredicateTreAnalyzerTest, require_that_minfeature_is_min_for_or) {
    auto slime(Builder().or_node
               ({Builder().and_node
                       ({Builder().feature("foo").value("bar"),
                         Builder().feature("baz").value("qux"),
                         Builder().feature("quux").value("corge")}),
                 Builder().and_node
                       ({Builder().feature("grault").value("garply"),
                         Builder().feature("waldo").value("fred")})})
               .build());
    PredicateTreeAnalyzer analyzer(slime->get());
    EXPECT_EQ(2, analyzer.getMinFeature());
    EXPECT_EQ(5, analyzer.getSize());
    EXPECT_EQ(5u, analyzer.getSizeMap().size());
    checkSizeMap(analyzer.getSizeMap(), "o0a0", 1);
    checkSizeMap(analyzer.getSizeMap(), "o0a1", 1);
    checkSizeMap(analyzer.getSizeMap(), "o0a2", 1);
    checkSizeMap(analyzer.getSizeMap(), "o1a0", 1);
    checkSizeMap(analyzer.getSizeMap(), "o1a1", 1);
}

TEST(PredicateTreAnalyzerTest, require_that_minfeature_rounds_up) {
    auto slime(Builder()
               .or_node({Builder().feature("foo").value("bar"),
                         Builder().feature("foo").value("bar"),
                         Builder().feature("foo").value("bar")}).build());
    PredicateTreeAnalyzer analyzer(slime->get());
    EXPECT_EQ(1, analyzer.getMinFeature());
    EXPECT_EQ(3, analyzer.getSize());
}

TEST(PredicateTreAnalyzerTest, require_that_multivalue_feature_set_considers_all_values) {
    {
        auto slime(Builder()
               .and_node({Builder().feature("foo").value("A").value("B"),
                          Builder().feature("foo").value("B")}).build());
        PredicateTreeAnalyzer analyzer(slime->get());
        EXPECT_EQ(1, analyzer.getMinFeature());
        EXPECT_EQ(2, analyzer.getSize());
    }
    {
        auto slime(Builder()
               .and_node({Builder().feature("foo").value("A").value("B"),
                          Builder().feature("foo").value("C")}).build());
        PredicateTreeAnalyzer analyzer(slime->get());
        EXPECT_EQ(2, analyzer.getMinFeature());
        EXPECT_EQ(2, analyzer.getSize());
    }
}

TEST(PredicateTreAnalyzerTest, require_that_not_features_dont_count_towards_minfeature_calculation) {
    auto slime(Builder()
               .and_node({Builder().feature("foo").value("A"),
                          Builder().neg().feature("foo").value("A"),
                          Builder().neg().feature("foo").value("B"),
                          Builder().feature("foo").value("B")}).build());
    PredicateTreeAnalyzer analyzer(slime->get());
    EXPECT_EQ(3, analyzer.getMinFeature());
    EXPECT_EQ(6, analyzer.getSize());
}

TEST(PredicateTreAnalyzerTest, require_that_not_ranges_dont_count_towards_minfeature_calculation) {
    auto slime(Builder()
               .and_node({Builder().feature("foo").range(0, 10),
                          Builder().neg().feature("foo").range(0, 10),
                          Builder().neg().feature("bar").range(0, 10),
                          Builder().feature("bar").range(0, 10)}).build());
    PredicateTreeAnalyzer analyzer(slime->get());
    EXPECT_EQ(3, analyzer.getMinFeature());
    EXPECT_EQ(6, analyzer.getSize());
}

TEST(PredicateTreAnalyzerTest, require_that_multilevel_AND_stores_sizes) {
    auto slime(Builder().and_node
               ({Builder().and_node
                       ({Builder().feature("foo").value("bar"),
                         Builder().feature("baz").value("qux"),
                         Builder().feature("quux").value("corge")}),
                 Builder().and_node
                       ({Builder().feature("grault").value("garply"),
                         Builder().feature("waldo").value("fred")})})
               .build());
    PredicateTreeAnalyzer analyzer(slime->get());
    EXPECT_EQ(5, analyzer.getMinFeature());
    EXPECT_EQ(5, analyzer.getSize());
    EXPECT_EQ(7u, analyzer.getSizeMap().size());
    checkSizeMap(analyzer.getSizeMap(), "a0", 3);
    checkSizeMap(analyzer.getSizeMap(), "a1", 2);
    checkSizeMap(analyzer.getSizeMap(), "a0a0", 1);
    checkSizeMap(analyzer.getSizeMap(), "a0a1", 1);
    checkSizeMap(analyzer.getSizeMap(), "a0a2", 1);
    checkSizeMap(analyzer.getSizeMap(), "a1a0", 1);
    checkSizeMap(analyzer.getSizeMap(), "a1a1", 1);
}

}  // namespace
